// @aztec/bb.js/dest/node-cjs/cbind/generated/sync.js
"use strict";
// AUTOGENERATED FILE - DO NOT EDIT
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncApi = void 0;
const msgpackr_1 = require("msgpackr");
const api_types_js_1 = require("./api_types.js");
function msgpackCall(wasm, cbind, input) { const inputBuffer = new msgpackr_1.Encoder({ useRecords: false }).pack(input); const encodedResult = wasm.cbindCall(cbind, inputBuffer); return new msgpackr_1.Decoder({ useRecords: false }).unpack(encodedResult); }
class SyncApi {
    constructor(wasm) {
        this.wasm = wasm;
    }
    circuitProve(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitProve)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["CircuitProve", msgpackCommand]]);
        if (variantName !== 'CircuitProveResponse') {
            throw new Error(`Expected variant name 'CircuitProveResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitProveResponse)(result);
    }
    circuitComputeVk(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitComputeVk)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["CircuitComputeVk", msgpackCommand]]);
        if (variantName !== 'CircuitComputeVkResponse') {
            throw new Error(`Expected variant name 'CircuitComputeVkResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitComputeVkResponse)(result);
    }
    circuitStats(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitStats)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["CircuitStats", msgpackCommand]]);
        if (variantName !== 'CircuitInfoResponse') {
            throw new Error(`Expected variant name 'CircuitInfoResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitInfoResponse)(result);
    }
    circuitVerify(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitVerify)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["CircuitVerify", msgpackCommand]]);
        if (variantName !== 'CircuitVerifyResponse') {
            throw new Error(`Expected variant name 'CircuitVerifyResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitVerifyResponse)(result);
    }
    clientIvcComputeStandaloneVk(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcComputeStandaloneVk)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["ClientIvcComputeStandaloneVk", msgpackCommand]]);
        if (variantName !== 'ClientIvcComputeStandaloneVkResponse') {
            throw new Error(`Expected variant name 'ClientIvcComputeStandaloneVkResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcComputeStandaloneVkResponse)(result);
    }
    clientIvcComputeIvcVk(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcComputeIvcVk)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["ClientIvcComputeIvcVk", msgpackCommand]]);
        if (variantName !== 'ClientIvcComputeIvcVkResponse') {
            throw new Error(`Expected variant name 'ClientIvcComputeIvcVkResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcComputeIvcVkResponse)(result);
    }
    clientIvcStart(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcStart)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["ClientIvcStart", msgpackCommand]]);
        if (variantName !== 'ClientIvcStartResponse') {
            throw new Error(`Expected variant name 'ClientIvcStartResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcStartResponse)(result);
    }
    clientIvcLoad(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcLoad)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["ClientIvcLoad", msgpackCommand]]);
        if (variantName !== 'ClientIvcLoadResponse') {
            throw new Error(`Expected variant name 'ClientIvcLoadResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcLoadResponse)(result);
    }
    clientIvcAccumulate(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcAccumulate)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["ClientIvcAccumulate", msgpackCommand]]);
        if (variantName !== 'ClientIvcAccumulateResponse') {
            throw new Error(`Expected variant name 'ClientIvcAccumulateResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcAccumulateResponse)(result);
    }
    clientIvcProve(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcProve)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["ClientIvcProve", msgpackCommand]]);
        if (variantName !== 'ClientIvcProveResponse') {
            throw new Error(`Expected variant name 'ClientIvcProveResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcProveResponse)(result);
    }
    clientIvcVerify(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcVerify)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["ClientIvcVerify", msgpackCommand]]);
        if (variantName !== 'ClientIvcVerifyResponse') {
            throw new Error(`Expected variant name 'ClientIvcVerifyResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcVerifyResponse)(result);
    }
    vkAsFields(command) {
        const msgpackCommand = (0, api_types_js_1.fromVkAsFields)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["VkAsFields", msgpackCommand]]);
        if (variantName !== 'VkAsFieldsResponse') {
            throw new Error(`Expected variant name 'VkAsFieldsResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toVkAsFieldsResponse)(result);
    }
    circuitWriteSolidityVerifier(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitWriteSolidityVerifier)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["CircuitWriteSolidityVerifier", msgpackCommand]]);
        if (variantName !== 'CircuitWriteSolidityVerifierResponse') {
            throw new Error(`Expected variant name 'CircuitWriteSolidityVerifierResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitWriteSolidityVerifierResponse)(result);
    }
    clientIvcCheckPrecomputedVk(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcCheckPrecomputedVk)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["ClientIvcCheckPrecomputedVk", msgpackCommand]]);
        if (variantName !== 'ClientIvcCheckPrecomputedVkResponse') {
            throw new Error(`Expected variant name 'ClientIvcCheckPrecomputedVkResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcCheckPrecomputedVkResponse)(result);
    }
    clientIvcStats(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcStats)(command);
        const [variantName, result] = msgpackCall(this.wasm, 'bbapi', [["ClientIvcStats", msgpackCommand]]);
        if (variantName !== 'ClientIvcStatsResponse') {
            throw new Error(`Expected variant name 'ClientIvcStatsResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcStatsResponse)(result);
    }
    destroy() {
        return this.wasm.destroy();
    }
}
exports.SyncApi = SyncApi;
